### PR TITLE
Add support for CASCADE in drop_view and update_view

### PR DESCRIPTION
**Pull request by jalada**
@pixielabs <3 Scenic!

We use lots of materialized views with dependencies and it can be quite painful to update them in a migration, especially remembering to recreate indexes.

This PR:

1. Adds an optional `cascade: true` to `drop_view` and `update_view`.
2. Makes `update_view` recreate any views that are lost during the view being updated, including any indexes on those views.
3. Stops the user from rolling back `drop_view` if they use `cascade: true`, because you can't tell what might've been lost. I've made the necessary change in `CommandRecorder`.

Things of note / bits I'm unsure about:

1. I don't know how this impacts the sqlite adapter, if at all.
2. I made `try_index_create` public because I'm using it to recreate indexes, but I'm not using it via the `on()` method. This is a new use of `IndexReapplication` and I'm conscious of that; perhaps some bigger API changes could be done here, but I didn't want to mess around too much!

If there's anything that needs changing, or any questions, let me know!

**December 2020 update**: Phew, the years just fly by don't they. There's a [summary of outstanding known issues](https://github.com/scenic-views/scenic/pull/218#issuecomment-736713334) at the bottom of this thread.

### CODE REVIEW

**Comment by jalada** [comment]
Oh, this is related to https://github.com/thoughtbot/scenic/issues/96

**Comment by jalada** [comment]
I've found a bug where this isn't working properly if you're updating a materialized view. Will push a fix!

**Comment by jalada** [comment]
Hound violations aside, I'd appreciate some suggestions on how to refactor the duplicated code, and also my use of `eval` in the tests to create similar migrations (one set materialized, one set not).

**Comment by jalada** [comment]
We've found an issue with the order in which views are captured and recreated; they are not in dependency order. Does anyone know of a way of asking for PSQL to return you views in order of dependency, or get the dependency tree for a bunch of views?

**Comment by annarankin** [comment]
Hello folks! This feature looks like a great addition 👍  Based on [this article](http://mwenus.blogspot.com/2014/04/db-view-dependencies-in-postgresql.html), I took a stab at a query to find PSQL views with their dependency order:

```sql
WITH RECURSIVE view_dependency_tree(parent_schema, parent_obj, child_schema, child_obj, ind, ord) AS 
(
  SELECT 
    vtu_parent.view_schema, 
    vtu_parent.view_name, 
    vtu_parent.table_schema, 
    vtu_parent.table_name, 
    '', 
    array[row_number() over (ORDER BY view_schema, view_name)]
  FROM information_schema.view_table_usage vtu_parent
  WHERE vtu_parent.view_schema = 'public'
  
  UNION ALL

  SELECT 
    vtu_child.view_schema, 
    vtu_child.view_name, 
    vtu_child.table_schema, 
    vtu_child.table_name, 
    vtu_parent.ind || '  ', 
    vtu_parent.ord || (row_number() over (ORDER BY view_schema, view_name))
  FROM view_dependency_tree vtu_parent, information_schema.view_table_usage vtu_child
  WHERE vtu_child.view_schema = vtu_parent.child_schema 
    AND vtu_child.view_name = vtu_parent.child_obj
) 
SELECT
  tree.parent_obj AS parent,
  tree.child_obj AS child,
  tree.ord
FROM view_dependency_tree tree
INNER JOIN information_schema.views view
  ON view.table_name = tree.child_obj
ORDER BY ord;
```
It seems to work for my views locally, hope it's helpful!

**Comment by jalada** [comment]
@annarankin this looks super useful! I’ll see if I can get it wired in

**Comment by jalada** [comment]
I had a problem with the `information_schema`-based query above, which is that it doesn't include materialized views. It only showed me dependencies between my views and the underlying tables.

This gives me a set of views (materialized or otherwise) and the 'level' on the tree they are on. This could be combined with our 'what views disappeared?' code to recreate them in the correct order:

```SQL
WITH RECURSIVE t AS
  ( SELECT c.oid,
           c.relname,
           0 AS LEVEL
   FROM pg_class c
   JOIN pg_namespace ON c.relnamespace = pg_namespace.oid
   WHERE c.relkind IN ('v', 'm') AND pg_namespace.nspname = 'public'
   UNION ALL SELECT c.oid,
                    c.relname,
                    a.level+1
   FROM t a
   JOIN pg_depend d ON d.refobjid=a.oid
   JOIN pg_rewrite w ON w.oid= d.objid AND w.ev_class!=a.oid
   JOIN pg_class c ON c.oid=w.ev_class)
SELECT relname, MAX(level) AS level
FROM t
GROUP BY relname;
```

It's slow, about 3s on my DB. It's also PSQL specific, so we'd still need some other solution for other versions of SQL that have dependent non-materialized views.

**Comment by annarankin** [comment]
Nice! Just for some qualitative feedback, I tried it on my local; it's pretty quick (~6ms) for 47 views and a max level of 5.

**Comment by jalada** [comment]
That's a lot of dependent views! That must've been frustrating without cascade support...

Another issue we've encountered: If we have dependent views with a `SELECT *` they don't pick up newly added fields in their parent views because PSQL expands out the `*` when it returns the view definition. Not sure what to do about that.

**Comment by ericfirth** [comment]
whats the status of this? we have the same problem at my company. If this is going to go in, I'd be happy to work on fixing the hound/code climate violations, but I wouldn't want to waste that time if this won't be going in scenic

File: lib/scenic/adapters/postgres.rb [hunk 1]
```diff
@@ -75,11 +75,25 @@ def create_view(name, sql_definition)
...
       # @param name The name of the view to update
       # @param sql_definition The SQL schema for the updated view.
+      # @param cascade Whether to drop and recreate dependent objects or not
       #
       # @return [void]
-      def update_view(name, sql_definition)
-        drop_view(name)
+      def update_view(name, sql_definition, cascade=false)
+        if cascade
+          # Get existing views that could be dependent on this one.
```
Comment (calebthompson) [inline lib/scenic/adapters/postgres.rb]:
```suggestion
```

File: lib/scenic/adapters/postgres.rb [hunk 2]
```diff
@@ -75,11 +75,25 @@ def create_view(name, sql_definition)
...
+          existing_views = views.reject{|v| v.name == name}
+
+          # Get indexes of existing materialized views
```
Comment (calebthompson) [inline lib/scenic/adapters/postgres.rb]:
```suggestion
```

File: lib/scenic/adapters/postgres.rb [hunk 3]
```diff
@@ -75,11 +75,25 @@ def create_view(name, sql_definition)
...
       # @param sql_definition The SQL schema for the updated view.
+      # @param cascade Whether to drop and recreate dependent objects or not
       #
       # @return [void]
-      def update_view(name, sql_definition)
-        drop_view(name)
+      def update_view(name, sql_definition, cascade=false)
+        if cascade
+          # Get existing views that could be dependent on this one.
+          existing_views = views.reject{|v| v.name == name}
```
Comment (calebthompson) [inline lib/scenic/adapters/postgres.rb]:
```suggestion
          potential_dependent_views = views.reject{|v| v.name == name}
```

File: lib/scenic/adapters/postgres.rb [hunk 4]
```diff
@@ -75,11 +75,25 @@ def create_view(name, sql_definition)
...
+
+          # Get indexes of existing materialized views 
+          indexes = Indexes.new(connection: connection)
+          view_indexes = existing_views.select(&:materialized).flat_map do |view|
```
Comment (calebthompson) [inline lib/scenic/adapters/postgres.rb]:
```suggestion
          potential_dependent_view_indexes = potential_dependent_views.select(&:materialized).flat_map do |view|
```

File: lib/scenic/adapters/postgres.rb [hunk 5]
```diff
@@ -75,11 +75,25 @@ def create_view(name, sql_definition)
...
+            indexes.on(view.name)
+          end
+        end
+
+        drop_view(name, cascade)
         create_view(name, sql_definition)
+
+        recreate_dropped_views(existing_views, views, view_indexes) if cascade
```
Comment (calebthompson) [inline lib/scenic/adapters/postgres.rb]:
```suggestion
        if cascade
          recreate_dropped_views(potential_dependent_views, views, potential_dependent_view_indexes) 
        end
```

File: lib/scenic/adapters/postgres.rb [hunk 6]
```diff
@@ -112,10 +126,11 @@ def replace_view(name, sql_definition)
       # This is typically called in a migration via {Statements#drop_view}.
       #
       # @param name The name of the view to drop
+      # @param cascade Whether to drop dependent objects or not
       #
       # @return [void]
-      def drop_view(name)
-        execute "DROP VIEW #{quote_table_name(name)};"
+      def drop_view(name, cascade=false)
+        execute "DROP VIEW #{quote_table_name(name)}#{" CASCADE" if cascade};"
```
[the remark this answers is not in the document]
Comment (calebthompson) [reply lib/scenic/adapters/postgres.rb]:
```suggestion
        execute "DROP VIEW #{quote_table_name(name)}#{' CASCADE' if cascade};"
```

File: lib/scenic/adapters/postgres.rb [hunk 7]
```diff
@@ -112,10 +126,11 @@ def replace_view(name, sql_definition)
       # This is typically called in a migration via {Statements#drop_view}.
       #
       # @param name The name of the view to drop
+      # @param cascade Whether to drop dependent objects or not
       #
       # @return [void]
-      def drop_view(name)
-        execute "DROP VIEW #{quote_table_name(name)};"
+      def drop_view(name, cascade=false)
```
Comment (calebthompson) [inline lib/scenic/adapters/postgres.rb]:
```suggestion
      def drop_view(name, cascade = false)
```

File: lib/scenic/adapters/postgres.rb [hunk 8]
```diff
@@ -144,32 +159,47 @@ def create_materialized_view(name, sql_definition)
       #
       # @param name The name of the view to update
       # @param sql_definition The SQL schema for the updated view.
+      # @param cascade Whether to drop and recreate dependent objects or not
```
Comment (calebthompson) [inline lib/scenic/adapters/postgres.rb]:
```suggestion
      # @param cascade Whether to drop and recreate dependent objects
```

File: lib/scenic/adapters/postgres.rb [hunk 9]
```diff
@@ -144,32 +159,47 @@ def create_materialized_view(name, sql_definition)
...
       #
       # @raise [MaterializedViewsNotSupportedError] if the version of Postgres
       #   in use does not support materialized views.
       #
       # @return [void]
-      def update_materialized_view(name, sql_definition)
+      def update_materialized_view(name, sql_definition, cascade=false)
```
[the remark this answers is not in the document]
Comment (calebthompson) [reply lib/scenic/adapters/postgres.rb]:
```suggestion
      def update_materialized_view(name, sql_definition, cascade = false)
```

File: lib/scenic/adapters/postgres.rb [hunk 10]
```diff
@@ -144,32 +159,47 @@ def create_materialized_view(name, sql_definition)
...
         raise_unless_materialized_views_supported
 
+        if cascade
+          # Get existing views that could be dependent on this one.
+          existing_views = views.reject{|v| v.name == name}
```
[the remark this answers is not in the document]
Comment (calebthompson) [reply lib/scenic/adapters/postgres.rb]:
```suggestion
          existing_views = views.reject { |v| v.name == name }
```

File: lib/scenic/adapters/postgres.rb [hunk 11]
```diff
@@ -144,32 +159,47 @@ def create_materialized_view(name, sql_definition)
...
+
+          # Get indexes of existing materialized views
```
[the remark this answers is not in the document]
Comment (calebthompson) [reply lib/scenic/adapters/postgres.rb]:
```suggestion
          # Get indexes of existing materialized views
```

File: lib/scenic/adapters/postgres.rb [hunk 12]
```diff
@@ -144,32 +159,47 @@ def create_materialized_view(name, sql_definition)
...
       # @raise [MaterializedViewsNotSupportedError] if the version of Postgres
       #   in use does not support materialized views.
       #
       # @return [void]
-      def update_materialized_view(name, sql_definition)
+      def update_materialized_view(name, sql_definition, cascade=false)
         raise_unless_materialized_views_supported
 
+        if cascade
+          # Get existing views that could be dependent on this one.
```
Comment (calebthompson) [inline lib/scenic/adapters/postgres.rb]:
```suggestion
```

File: lib/scenic/adapters/postgres.rb (hunk 10 above)
[the remark this answers is not in the document]
Comment (calebthompson) [reply lib/scenic/adapters/postgres.rb]:
```suggestion
          potential_dependent_views = views.reject{|v| v.name == name}
```

File: lib/scenic/adapters/postgres.rb (hunk 11 above)
Comment (calebthompson) [inline lib/scenic/adapters/postgres.rb]:
```suggestion
```

File: lib/scenic/adapters/postgres.rb [hunk 13]
```diff
@@ -144,32 +159,47 @@ def create_materialized_view(name, sql_definition)
...
+          existing_views = views.reject{|v| v.name == name}
+
+          # Get indexes of existing materialized views 
+          indexes = Indexes.new(connection: connection)
+          view_indexes = existing_views.select(&:materialized).flat_map do |view|
```
[the remark this answers is not in the document]
Comment (calebthompson) [reply lib/scenic/adapters/postgres.rb]:
```suggestion
          potential_dependent_view_indexes = potential_dependent_views.
            select(&:materialized).
            flat_map { |view| indexes.on(view.name) }
```

File: lib/scenic/adapters/postgres.rb [hunk 14]
```diff
@@ -144,32 +159,47 @@ def create_materialized_view(name, sql_definition)
...
+            indexes.on(view.name)
```
Comment (calebthompson) [inline lib/scenic/adapters/postgres.rb]:
```suggestion
```

File: lib/scenic/adapters/postgres.rb [hunk 15]
```diff
@@ -144,32 +159,47 @@ def create_materialized_view(name, sql_definition)
...
+          end
```
Comment (calebthompson) [inline lib/scenic/adapters/postgres.rb]:
```suggestion
```

File: lib/scenic/adapters/postgres.rb [hunk 16]
```diff
@@ -144,32 +159,47 @@ def create_materialized_view(name, sql_definition)
...
+        end
+
         IndexReapplication.new(connection: connection).on(name) do
-          drop_materialized_view(name)
+          drop_materialized_view(name, cascade)
           create_materialized_view(name, sql_definition)
         end
+
+        recreate_dropped_views(existing_views, views, view_indexes) if cascade
```
Comment (calebthompson) [inline lib/scenic/adapters/postgres.rb]:
```suggestion
        if cascade
          recreate_dropped_views(potential_dependent_views, views, potential_dependent_view_indexes)
        end
```

File: lib/scenic/adapters/postgres.rb [hunk 17]
```diff
@@ -144,32 +159,47 @@ def create_materialized_view(name, sql_definition)
...
       # This is typically called in a migration via {Statements#update_view}.
       #
       # @param name The name of the materialized view to drop.
+      # @param cascade Whether to drop dependent objects or not.
       # @raise [MaterializedViewsNotSupportedError] if the version of Postgres
       #   in use does not support materialized views.
       #
       # @return [void]
-      def drop_materialized_view(name)
+      def drop_materialized_view(name, cascade=false)
```
[the remark this answers is not in the document]
Comment (calebthompson) [reply lib/scenic/adapters/postgres.rb]:
```suggestion
      def drop_materialized_view(name, cascade = false)
```

File: lib/scenic/adapters/postgres.rb [hunk 18]
```diff
@@ -144,32 +159,47 @@ def create_materialized_view(name, sql_definition)
...
+
+        recreate_dropped_views(existing_views, views, view_indexes) if cascade
       end
 
       # Drops a materialized view in the database
       #
       # This is typically called in a migration via {Statements#update_view}.
       #
       # @param name The name of the materialized view to drop.
+      # @param cascade Whether to drop dependent objects or not.
```
Comment (calebthompson) [inline lib/scenic/adapters/postgres.rb]:
```suggestion
      # @param cascade Whether to drop dependent objects.
```

File: lib/scenic/adapters/postgres.rb [hunk 19]
```diff
@@ -144,32 +159,47 @@ def create_materialized_view(name, sql_definition)
...
       # @raise [MaterializedViewsNotSupportedError] if the version of Postgres
       #   in use does not support materialized views.
       #
       # @return [void]
-      def drop_materialized_view(name)
+      def drop_materialized_view(name, cascade=false)
         raise_unless_materialized_views_supported
-        execute "DROP MATERIALIZED VIEW #{quote_table_name(name)};"
+        execute "DROP MATERIALIZED VIEW #{quote_table_name(name)}#{" CASCADE" if cascade};"
```
[the remark this answers is not in the document]
Comment (calebthompson) [reply lib/scenic/adapters/postgres.rb]:
```suggestion
        execute "DROP MATERIALIZED VIEW #{quote_table_name(name)}#{' CASCADE' if cascade};"
```

File: lib/scenic/adapters/postgres.rb [hunk 20]
```diff
@@ -238,6 +268,24 @@ def refresh_dependencies_for(name)
           connection,
         )
       end
+
+      def recreate_dropped_views(old_views, current_views, indexes=[])
```
[the remark this answers is not in the document]
Comment (calebthompson) [reply lib/scenic/adapters/postgres.rb]:
```suggestion
      def recreate_dropped_views(old_views, current_views, indexes = [])
```

File: lib/scenic/adapters/postgres.rb [hunk 21]
```diff
@@ -238,6 +268,24 @@ def refresh_dependencies_for(name)
...
+        index_reapplier = IndexReapplication.new(connection: connection)
+
+        # Find any views that were lost
+        dropped_views = old_views.reject{|ov| current_views.any?{|cv| ov.name == cv.name}}
```
[the remark this answers is not in the document]
Comment (calebthompson) [reply lib/scenic/adapters/postgres.rb]:
```suggestion
        dropped_views = old_views.reject do |ov|
          current_views.any? { |cv| ov.name == cv.name }
       end
```

File: lib/scenic/adapters/postgres.rb [hunk 22]
```diff
@@ -238,6 +268,24 @@ def refresh_dependencies_for(name)
...
+        # Recreate them
+        dropped_views.each do |view|
+          if view.materialized
+            create_materialized_view view.name, view.definition
+            # Also recreate any indexes that were lost
+            lost_indexes = indexes.select{|index| index.object_name == view.name}
```
[the remark this answers is not in the document]
Comment (calebthompson) [reply lib/scenic/adapters/postgres.rb]:
```suggestion
            lost_indexes = indexes.select { |index| index.object_name == view.name }
```

File: lib/scenic/adapters/postgres.rb [hunk 23]
```diff
@@ -238,6 +268,24 @@ def refresh_dependencies_for(name)
...
+      def recreate_dropped_views(old_views, current_views, indexes=[])
+        index_reapplier = IndexReapplication.new(connection: connection)
+
+        # Find any views that were lost
+        dropped_views = old_views.reject{|ov| current_views.any?{|cv| ov.name == cv.name}}
+        # Recreate them
+        dropped_views.each do |view|
+          if view.materialized
+            create_materialized_view view.name, view.definition
+            # Also recreate any indexes that were lost
```
Comment (calebthompson) [inline lib/scenic/adapters/postgres.rb]:
```suggestion
```

File: lib/scenic/adapters/postgres.rb [hunk 24]
```diff
@@ -238,6 +268,24 @@ def refresh_dependencies_for(name)
           connection,
         )
       end
+
+      def recreate_dropped_views(old_views, current_views, indexes=[])
+        index_reapplier = IndexReapplication.new(connection: connection)
+
+        # Find any views that were lost
+        dropped_views = old_views.reject{|ov| current_views.any?{|cv| ov.name == cv.name}}
+        # Recreate them
```
Comment (calebthompson) [inline lib/scenic/adapters/postgres.rb]:
```suggestion
```

File: lib/scenic/adapters/postgres.rb [hunk 25]
```diff
@@ -238,6 +268,24 @@ def refresh_dependencies_for(name)
           connection,
         )
       end
+
+      def recreate_dropped_views(old_views, current_views, indexes=[])
+        index_reapplier = IndexReapplication.new(connection: connection)
+
+        # Find any views that were lost
```
Comment (calebthompson) [inline lib/scenic/adapters/postgres.rb]:
```suggestion
```

File: lib/scenic/statements.rb [hunk 26]
```diff
@@ -50,16 +50,18 @@ def create_view(name, version: nil, sql_definition: nil, materialized: false)
     #   `version` argument to {#create_view}.
     # @param materialized [Boolean] Set to true if dropping a meterialized view.
     #   defaults to false.
+    # @param cascade [Boolean] Set to true if dependent objects should also be
+    #   dropped. defaults to false.
```
Comment (calebthompson) [inline lib/scenic/statements.rb]:
```suggestion
    #   dropped. Defaults to false.
```

File: lib/scenic/statements.rb [hunk 27]
```diff
@@ -50,16 +50,18 @@ def create_view(name, version: nil, sql_definition: nil, materialized: false)
     #   `version` argument to {#create_view}.
     # @param materialized [Boolean] Set to true if dropping a meterialized view.
     #   defaults to false.
```
Comment (calebthompson) [inline lib/scenic/statements.rb]:
```suggestion
    #   Defaults to false.
```

File: lib/scenic/statements.rb [hunk 28]
```diff
@@ -77,12 +79,14 @@ def drop_view(name, revert_to_version: nil, materialized: false)
     #   `rake db rollback`
     # @param materialized [Boolean] True if updating a materialized view.
     #   Defaults to false.
+    # @param cascade [Boolean] Set to true if dependent objects should also be
+    #   dropped. defaults to false.
```
Comment (calebthompson) [inline lib/scenic/statements.rb]:
```suggestion
    #   dropped. Defaults to false.
```

File: spec/integration/cascade_spec.rb [hunk 29]
```diff
@@ -0,0 +1,99 @@
...
+
+  def index_migration
+    Class.new(migration_class) do
+      def change
+        add_index :dependent_greetings, :greeting
+      end
+    end
+  end
+
+
```
[the remark this answers is not in the document]
Comment (calebthompson) [reply spec/integration/cascade_spec.rb]:
```suggestion
```

File: spec/integration/cascade_spec.rb [hunk 30]
```diff
@@ -0,0 +1,120 @@
...
+  end
+
+  def index_migration
+    Class.new(migration_class) do
+      def change
+        add_index :dependent_greetings, :greeting
+      end
+    end
+  end
+
```
Comment (calebthompson) [inline spec/integration/cascade_spec.rb]:
```suggestion
```

File: spec/integration/revert_spec.rb [hunk 31]
```diff
@@ -52,23 +52,4 @@ def change
     end
   end
```
Comment (calebthompson) [inline spec/integration/revert_spec.rb]:
```suggestion
```

File: spec/support/migration_helpers.rb [hunk 32]
```diff
@@ -0,0 +1,27 @@
+module MigrationHelpers
+
```
[the remark this answers is not in the document]
Comment (calebthompson) [reply spec/support/migration_helpers.rb]:
```suggestion
```

File: spec/support/migration_helpers.rb [hunk 33]
```diff
@@ -0,0 +1,27 @@
...
+  end
+
+  def execute(sql)
+    connection.execute(sql)
+  end
+
+  def connection
+    ActiveRecord::Base.connection
+  end
+
```
[the remark this answers is not in the document]
Comment (calebthompson) [reply spec/support/migration_helpers.rb]:
```suggestion
```

**Comment by dansteele** [comment]
Would be good to see this merged in 👍

**Comment by brnosouza** [comment]
Are there any plans to merge this?

**Comment by jtemplet** [comment]
Hey all, just wanted to follow up to see if there's any help needed to get this merged in?

cc @calebthompson

**Comment by jalada** [comment]
@jtemplet there are two problems with this PR that we identified, which makes it a bit more 'work-in-progress' than we originally thought:

1. [When capturing views for recreation, they're not in dependency order](https://github.com/scenic-views/scenic/pull/218#issuecomment-341363586). There's a couple of proposed solutions further down the thread. But one doesn't work, and one was slow & PSQL specific.
2. [Views with SELECT * in the definition](https://github.com/scenic-views/scenic/pull/218#issuecomment-344609640). This is pretty frustrating and I didn't know what to do about it.

So yes, it definitely needs some help, quite considerable help.

**Comment by derekprior** [comment]
We're having some issues with getting things in a stable, dependency order in a couple of different places in the codebase. I'm inclined to close this for now, but if we're able to confidently fix that issue then we could potentially revisit. See: https://github.com/scenic-views/scenic/labels/dependencies

**Comment by jalada** [comment]
I wonder if https://github.com/scenic-views/scenic/pull/398 would help when revisiting this problem.

Context: We're doing a Ruby 2.7 -> 3.0 upgrade and our old fork isn't compatible, so we're thinking again about how to solve this problem.

**Comment by 23tux** [comment]
I would also be interested in this feature, are there any plans on picking this up again?

**Pull request closed without merging**